File: CPP/data_structures/stacks/InfixToPostfix.cpp

/*
Infix to Postfix Conversion (Shunting Yard Algorithm)

Description:
This algorithm converts an infix expression (e.g., A+B*C) into a postfix expression (e.g., A B C * +)
using the C++ Standard Template Library (STL) stack.
It follows operator precedence and associativity rules to ensure correct order of operations.
The stack temporarily stores operators and parentheses during the conversion process.

Example:
Input:  A+B*C
Output: A B C * +

Time Complexity: O(n)
Each character in the infix expression is scanned once and pushed/popped from the stack at most once.

Space Complexity: O(n)
A stack is used to store operators and parentheses during processing.

Reference:
https://en.wikipedia.org/wiki/Shunting_yard_algorithm
*/
#include <iostream>
#include <stack>
#include <string>
using namespace std;

// Function to define operator precedence
int precedence(char op) {
    if (op == '+' || op == '-') return 1;
    if (op == '*' || op == '/') return 2;
    if (op == '^') return 3;
    return 0;
}

// Function to check if character is an operator
bool isOperator(char c) {
    return (c == '+' || c == '-' || c == '*' || c == '/' || c == '^');
}

// Function to convert infix to postfix
string infixToPostfix(string infix) {
    stack<char> st;
    string postfix = "";

    for (int i = 0; i < infix.length(); i++) {
        char c = infix[i];

        // If operand, add to output
        if (isalnum(c)) {
            postfix += c;
        }
        // If '(', push to stack
        else if (c == '(') {
            st.push(c);
        }
        // If ')', pop until '('
        else if (c == ')') {
            while (!st.empty() && st.top() != '(') {
                postfix += st.top();
                st.pop();
            }
            st.pop(); // remove '('
        }
        // If operator
        else if (isOperator(c)) {
            while (!st.empty() && precedence(st.top()) >= precedence(c)) {
                postfix += st.top();
                st.pop();
            }
            st.push(c);
        }
    }

    // Pop remaining operators
    while (!st.empty()) {
        postfix += st.top();
        st.pop();
    }

    return postfix;
}

int main() {
    string infix;
    cout << "Enter an infix expression: ";
    cin >> infix;

    string postfix = infixToPostfix(infix);
    cout << "Postfix expression: " << postfix << endl;

    return 0;
}

File: Java/data_structures/stacks/InfixToPostfix.java
/*
 * Algorithm Name
   Infix to Postfix Conversion

 * Programming Language
   Java

 * Category
   Data Structures / Stack Applications

 * Difficulty Level
   Medium

 * Algorithm Description:

   1. Problem it solves:
      -> Converts an arithmetic expression from Infix form (e.g., A + B * C)
         to Postfix form (e.g., A B C * +), which can be easily evaluated by computers.

   2. Approach / Idea:
      -> Use a stack to temporarily hold operators.
      -> Steps:
         - Scan the infix expression from left to right.
         - If the character is an operand (A–Z, a–z, 0–9), add it to the output.
         - If the character is '(', push it to stack.
         - If the character is ')', pop and output until '(' is found.
         - If the character is an operator (+, -, *, /, ^):
             * Pop operators from stack to output until the stack is empty or
               top of stack has lower precedence.
             * Then push the current operator to the stack.
         - After the full scan, pop and output all remaining operators from stack.

   3. Complexity:
      Time: O(n) — each element is pushed and popped at most once.
      Space: O(n) — stack and output storage.

 * Author: Pradyumn Pratap Singh (Strange)
 */

package DSA_Code.Java.data_structures.stacks;

import java.util.Stack;

public class InfixToPostfix {

    // Function to return precedence of operators
    private static int precedence(char ch) {
        switch (ch) {
            case '+':
            case '-':
                return 1;
            case '*':
            case '/':
                return 2;
            case '^':
                return 3;
        }
        return -1;
    }

    // Function to convert infix expression to postfix
    public static String infixToPostfix(String expression) {
        StringBuilder result = new StringBuilder();
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < expression.length(); i++) {
            char c = expression.charAt(i);

            // If the scanned character is an operand, add it to output
            if (Character.isLetterOrDigit(c)) {
                result.append(c);
            }
            // If the scanned character is '(', push it to stack
            else if (c == '(') {
                stack.push(c);
            }
            // If the scanned character is ')', pop and output until '(' is found
            else if (c == ')') {
                while (!stack.isEmpty() && stack.peek() != '(') {
                    result.append(stack.pop());
                }
                if (!stack.isEmpty() && stack.peek() == '(') {
                    stack.pop();
                } else {
                    System.out.println("Invalid Expression: Mismatched parentheses.");
                    return "";
                }
            }
            // If an operator is encountered
            else {
                while (!stack.isEmpty() && precedence(c) <= precedence(stack.peek())) {
                    result.append(stack.pop());
                }
                stack.push(c);
            }
        }

        // Pop all remaining operators from the stack
        while (!stack.isEmpty()) {
            if (stack.peek() == '(') {
                System.out.println("Invalid Expression: Mismatched parentheses.");
                return "";
            }
            result.append(stack.pop());
        }

        return result.toString();
    }

    // Driver code for testing
    public static void main(String[] args) {
        String infixExp1 = "A+B*C";
        String infixExp2 = "(A+B)*C";
        String infixExp3 = "A+B*(C^D-E)^(F+G*H)-I";

        System.out.println("Infix Expression 1: " + infixExp1);
        System.out.println("Postfix Expression 1: " + infixToPostfix(infixExp1));

        System.out.println("\nInfix Expression 2: " + infixExp2);
        System.out.println("Postfix Expression 2: " + infixToPostfix(infixExp2));

        System.out.println("\nInfix Expression 3: " + infixExp3);
        System.out.println("Postfix Expression 3: " + infixToPostfix(infixExp3));
    }
}
